### PR TITLE
Add badges to Advanced Search for cookbooks

### LINKED ISSUE #1177: There should be a search option for Chef Partner Cookbooks
It would be very helpful to be able to filter the search options to find Chef Partner Cookbooks. This should be available as a fixed URL, so people can link to the listing.

In support of ROAD-73 (Display quality metrics for each cookbook on the Supermarket)
screenshot 2015-12-03 15 43 28

### PR DESCRIPTION
**Pull request by robbkidd**
Closes #1177 

Mimicks searching for cookbooks by supported platform.

- [x] Feature spec(s) added for searching by badge

### CODE REVIEW

**Comment by robbkidd** [comment]
@nellshamrell @nathenharvey @mattray @jjasghar 
Search form submitted in screenshot below produces a linkable URL like:
```
/cookbooks?utf8=%E2%9C%93&q=&badges%5B%5D=partner&platforms%5B%5D=
```
screen shot 2016-01-29 at 10 04 14 am

**Comment by jjasghar** [comment]
image

BOOM! Thats a win.

File: app/models/concerns/badgeable.rb [hunk 1]
```diff
@@ -7,6 +7,17 @@ module Badgeable
...
   BADGES = %w(partner).freeze
 
+  module ClassMethods
+    #
+    # Provides a search method for Badgeable AR::Base models
+    #
+    def with_badges(badges)
+      badges = Array(badges).map(&:to_s)
+      search_mask = (badges & BADGES).map { |b| 2**BADGES.index(b) }.inject(0, :+)
+      where('badges_mask & ? > 0', search_mask)
```
Comment (robbkidd) [inline app/models/concerns/badgeable.rb]: I can foresee needing an index on cookbooks.badges_mask.

**Comment by robbkidd** [comment]
@nathenharvey said :shipit:

**Pull request merged**